### PR TITLE
feat: ability to compile lsp to wasm

### PR DESCRIPTION
**Pull request by lgalabru**
We're revisiting the architecture of the code and repo so we unblock https://github.com/hirosystems/clarity-lsp/issues/48. 
With this PR, we're extracting 4 sub-crates:
- clarinet-files: logic for dealing with the repository layout: Project manifests, Network manifests, Contracts. We're abstracting the underlying storage (FileSystem/URL/...) so we have 1) a clean way to deal with stuffs like VFS (required in VSCode) 2) make this layer compatible with WASM (anything related to filesystems make WASM builds incompatible)
- clarinet-deployments: core logic for building deployments
- clarity-lsp: building blocks for building a LSP. in the Clarinet crate, we're implementing a LSP (communication happening through Stdin/Stdout - abstracted with tower-lsp), but the new vscode extension would work using cross-webworker messaging.
- clarity-utils: some sort of potpourri.

This extraction is following the architecture suggested in https://github.com/hirosystems/clarinet/pull/406.

### CODE REVIEW

**Review by obycode — COMMENTED** [review]
This is a really nice change @lgalabru. Very clean! Just a couple of comments.

components/clarity-lsp/src/capabilities.rs is an empty file. Can it be deleted?

File: components/clarinet-deployments/src/types.rs [hunk 1]
```diff
@@ -68,17 +80,19 @@ pub struct RequirementPublishSpecificationFile {
     pub contract_id: String,
     pub remap_sender: String,
     pub remap_principals: Option<BTreeMap<String, String>>,
-    pub path: String,
     pub cost: u64,
+    #[serde(flatten)]
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply components/clarinet-deployments/src/types.rs:84]: So if I remember correctly:
- Default
```
{
   ...
   location: { Url: { url: "http://..." } } 
}
```
- Untag
```
{
   ...
   location: { url: "http://..." } 
}
```
- Untag + Flatten
```
{
   ...
   url: "http://..." 
}
```

File: src/generate/project.rs [hunk 2]
```diff
@@ -17,7 +17,7 @@ impl GetChangesForNewProject {
         }
     }
 
-    pub fn run(&mut self) -> Vec<Changes> {
+    pub fn run(&mut self) -> Result<Vec<Changes>, String> {
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/generate/project.rs:20]: Streamlining the interfaces... Technically we could test and return an error if there's a file collision, for instance?

File: components/clarinet-files/src/network_manifest.rs [hunk 3]
```diff
@@ -207,66 +209,60 @@ pub struct AccountConfig {
...
-        });
-        let chain_config = ChainConfig::from_path(&chain_config_path, networks);
-        chain_config
+    ) -> Result<NetworkManifest, String> {
+        let network_manifest_location =
+            project_manifest_location.get_network_manifest_location(&networks.1)?;
+        NetworkManifest::from_location(&network_manifest_location, networks)
     }
 
     #[allow(non_fmt_panics)]
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply components/clarinet-files/src/network_manifest.rs]: Addressed with https://github.com/hirosystems/clarinet/pull/413/commits/cdf4e6def986d43734a6dc0b7aff688b7f06f409

File: components/clarinet-files/src/project_manifest.rs [hunk 4]
```diff
@@ -30,27 +27,25 @@ pub struct ProjectConfigFile {
...
+#[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct ProjectManifest {
     pub project: ProjectConfig,
     #[serde(serialize_with = "toml::ser::tables_last")]
     pub contracts: BTreeMap<String, ContractConfig>,
     #[serde(rename = "repl")]
     pub repl_settings: repl::Settings,
-    #[serde(skip_serializing, skip_deserializing)]
-    pub path: PathBuf,
+    // #[serde(skip_serializing, skip_deserializing)]
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply components/clarinet-files/src/project_manifest.rs]: Addressed with https://github.com/hirosystems/clarinet/pull/413/commits/cdf4e6def986d43734a6dc0b7aff688b7f06f409

**Comment by lgalabru** [comment]
Thank you for the review on a Saturday morning @obycode!

File: components/clarinet-deployments/src/types.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (obycode) [reply components/clarinet-deployments/src/types.rs:84]: Oh, right. Thanks!

**Comment by obycode** [comment]
Thanks for the explanations and updates. Can components/clarity-lsp/src/capabilities.rs be deleted or is the empty file kept for some reason?

**Review by obycode — APPROVED** [review]
Lgtm!

**Pull request merged**